crepecake/src/main/java/com/google/cloud/tools/crepecake/registry/RegistryClient.java

```diff
@@ -0,0 +1,130 @@
+import com.google.cloud.tools.crepecake.json.JsonTemplateMapper;
+import com.google.cloud.tools.crepecake.registry.json.ErrorEntryTemplate;
+import com.google.cloud.tools.crepecake.registry.json.ErrorResponseTemplate;
+import java.io.IOException;
+import java.net.HttpURLConnection;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.nio.file.Path;
+import javax.annotation.Nullable;
+
+/** Interfaces with a registry. */
+public class RegistryClient {
+
+  // TODO: This should be configurable.
+  private static final String PROTOCOL = "http";
+
+  @Nullable private final Authorization authorization;
+  private final String serverUrl;
+  private final String imageName;
+
+  public RegistryClient(@Nullable Authorization authorization, String serverUrl, String imageName) {
+    this.authorization = authorization;
+    this.serverUrl = serverUrl;
+    this.imageName = imageName;
+  }
+
+  /** Pulls the image manifest for a specific tag. */
+  public ManifestTemplate pullManifest(String imageTag) throws IOException, RegistryException {
+    ManifestPuller manifestPuller = new ManifestPuller(imageTag);
+    return (ManifestTemplate) callRegistryEndpoint(null, manifestPuller);
+  }
+
+  /**
+   * Downloads the BLOB to a file.
+   *
+   * @param blobDigest the digest of the BLOB to download
+   * @param destPath the path of the file to write to
+   * @return a {@link Blob} backed by the file at {@code destPath}. The file at {@code destPath}
+   *     must exist for {@link Blob} to be valid.
+   */
+  public Blob pullBlob(DescriptorDigest blobDigest, Path destPath)
+      throws RegistryException, IOException {
+    BlobPuller blobPuller = new BlobPuller(blobDigest, destPath);
+    return (Blob) callRegistryEndpoint(null, blobPuller);
+  }
+
+  private URL getApiRoute(String routeSuffix) throws MalformedURLException {
+    String apiBase = "/v2/";
+    return new URL(PROTOCOL + "://" + serverUrl + apiBase + imageName + routeSuffix);
+  }
+
+  /**
+   * Calls the registry endpoint.
+   *
+   * @param url the endpoint URL to call, or {@code null} to use default from {@code
+   *     registryEndpointProvider}
+   * @param registryEndpointProvider the {@link RegistryEndpointProvider} to the endpoint
+   */
+  private Object callRegistryEndpoint(
+      @Nullable URL url, RegistryEndpointProvider registryEndpointProvider)
+      throws IOException, RegistryException {
+    if (url == null) {
+      url = getApiRoute(registryEndpointProvider.getApiRouteSuffix());
+    }
+
+    try (Connection connection = new Connection(url)) {
+      Request.Builder builder = Request.builder();
+      if (authorization != null) {
+        builder.setAuthorization(authorization);
+      }
+      Response response = connection.get(builder.build());
+
+      return registryEndpointProvider.handleResponse(response);
+
+    } catch (HttpResponseException ex) {
+      switch (ex.getStatusCode()) {
+        case HttpURLConnection.HTTP_BAD_REQUEST:
```

Comment:
depending on switch case fallthrough is extremely error prone.

Reply:
The throw at the end of the this statement group prevents the fallthrough from happening.

Reply:
It's not that it's used incorrectly now. It's that it will be used incorrectly in the future as the code evolves. For this reason most static analysis tools report this as an error. See Chapter 2 of van der Linden's Deep C Secrets for more context; but tldr don't use switch statements that depend on fallthrough.

Reply:
Hmm, it doesn't seem that this code qualifies as fallthrough since there is no code between the labels. I'm not sure what would be an alternative solution here since there's no way to combine labels into an OR without putting a colon-case in-between?

Reply:
You could use individual case statements or  better yet use if-else instead of switch.

Reply:
Agreed, this block might be more readable as if/else.